### ISSUE TITLE
Need an "Application" abstraction layer

### ISSUE THREAD

**Issue by dsblank** [tsileo/little-boxes#6]
First, thanks for creating the "little-boxes" spinoff for the abstraction of the "backend"! 

However, all of the "business logic" is actually in microblog.pub's [MicroblogPubBackend](https://github.com/tsileo/microblog.pub/blob/master/activitypub.py#L54). It seems that the logic there could be abstracted just slightly away from the database which would then allow activitypub to be largely implemented by providing an API to the database. Then the porting of a "backend" would be largely providing a database instance (MongoDB, SqliteDB, etc) to an Application instance. Something like an "Application" class could include the database instance, and could contain the "fetch" methods separate from the ActivityPub class itself.

Perhaps your goals can't include this. No problem. But I suspect that there are others interested in knowing if we should try to make little-boxes be that generic library.

**Comment by dsblank** [comment]
Edited above comment (in case you are reading this via email)

**Comment by tsileo** [comment]
Hey!

First, it was a huge chunk of work to start moving all the ActivityPub related bits to little-boxes. So my first goal was to get microblog.pub working with little-boxes, without too much modifications. And now this is done I will see how to improve the little-boxes API. So thanks for the feedback.

I disagree when you say all the "business logic" remains in microblog.pub backend, there still some parts that need to be moved (because it's still in super early stage). But for example, the logic that posts an accept in response to a follow is part of little-boxes. What remains in the microblog.pub backend are "side-effects": like "saving a new activity in the oubox", "someone liked something from the inbox". Every server might want to do something different, adds additional logic. So I kind of like the "backend of side-effect idea". An area I really want to improve, is helper for serving/building the different collection that needs to exposed by a server.

Also all the ActivityPub related helpers (RSA, JSON-LD signature, HTTPSig...) was made available without any tie to a web framework, and I had to re-implement all these helpers because I haven't found any working implementation that suited me. This alone I think has good value in the library.

One more point, with the "current codebase" it would be possible to implement an ActivityPub client (I have started working on something but it's not published yet), but what you're talking sounds more like a "backbone activitypub server", and that's something that could see implemented on top of little-boxes, but not part of it.

About the globals, I'm not too happy with them too, and I started moving more things inside the backend, like the debug mode (and for the backend, one is expected to call `ap.get_backend()` to get the currently activated backend).
I tried to find other ways, but I don't like the tradeoffs, so I think the backend will remain global for now (also there's no more DB in little-boxes?).

To conclude, as I said previously, I'm still not sure about the final scope of little-boxes, but I know that it will stay quite low-level, and I'm pretty sure if the API matures a little-bit, it could fit your purpose.

But I'm open to discussion, thanks for your feedback.
